src/libraries/Microsoft.Extensions.DependencyModel/src/LibraryCollectionExtensions.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    using System;
+    using System.Collections.Generic;
+
+    internal static class LibraryCollectionExtensions
+    {
+        public static Dictionary<string, T> LibraryCollectionToDictionary<T>(this IReadOnlyList<T> collection) where T : Library
+        {
+            var dictionary = new Dictionary<string, T>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var element in collection)
+            {
+                if (dictionary.ContainsKey(element.Name))
```

Comment:
I wonder if instead of having to check `ContainsKey` on every element, and then check again in `.Add`, if we just had a try-catch around the foreach, caught ArgumentException, and only then wrapped the exception with the `element.Name`. That way we wouldn't need to double-check for every element.

Reply:
I did a really quick benchmark of this:

```C#
    [MemoryDiagnoser(displayGenColumns: false)]
    public class CollectionBenchmarks
    {
        private IReadOnlyList<RuntimeLibrary> _libraries;

        public CollectionBenchmarks()
        {
            const string depsFilePath = @"C:\Users\eerhardt\source\repos\CollectionBenchmark\CollectionBenchmark\bin\Release\net462\CollectionBenchmark.deps.json";
            _libraries = new DependencyContextJsonReader().Read(File.OpenRead(depsFilePath)).RuntimeLibraries;
        }

        [Benchmark(Baseline = true)]
        public Dictionary<string, RuntimeLibrary> LibraryCollectionToDictionary1() => _libraries.LibraryCollectionToDictionary1();

        [Benchmark]
        public Dictionary<string, RuntimeLibrary> LibraryCollectionToDictionary2() => _libraries.LibraryCollectionToDictionary2();
    }

    internal static class MyExtensions
    {
        internal static Dictionary<string, T> LibraryCollectionToDictionary1<T>(this IReadOnlyList<T> collection) where T : Library
        {
            var dictionary = new Dictionary<string, T>(StringComparer.OrdinalIgnoreCase);

            foreach (var element in collection)
            {
                if (dictionary.ContainsKey(element.Name))
                {
                    throw new ArgumentException("duplicate");
                }

                dictionary.Add(element.Name, element);
            }

            return dictionary;
        }

        internal static Dictionary<string, T> LibraryCollectionToDictionary2<T>(this IReadOnlyList<T> collection) where T : Library
        {
            var dictionary = new Dictionary<string, T>(StringComparer.OrdinalIgnoreCase);

            foreach (var element in collection)
            {
                try
                {
                    dictionary.Add(element.Name, element);
                }
                catch (ArgumentException)
                {
                    throw new ArgumentException("duplicate");
                }
            }

            return dictionary;
        }
    }
```

The deps.json file I'm pointing to has 69 RuntimeLibraries in it, which is a relatively small .deps.json file.

With the following results:

|                         Method |     Mean |     Error |    StdDev | Ratio | Allocated |
|------------------------------- |---------:|----------:|----------:|------:|----------:|
| LibraryCollectionToDictionary1 | 8.436 us | 0.1090 us | 0.0910 us |  1.00 |      5 KB |
| LibraryCollectionToDictionary2 | 4.797 us | 0.0632 us | 0.0591 us |  0.57 |      5 KB |